contracts/contracts/coordination/Subscription.sol

```diff
@@ -0,0 +1,261 @@
+     * @param ritualId The ID of the ritual
+     * @param encryptor The address of the encryptor
+     * @return The key used to lookup authorizations
+     */
+    function lookupKey(uint32 ritualId, address encryptor) internal pure returns (bytes32) {
+        return keccak256(abi.encodePacked(ritualId, encryptor));
+    }
+
+    /**
+     * @notice Creates a new subscription
+     * @param ritualId The ID of the ritual
+     */
+    function newSubscription(uint32 ritualId) external {
+        uint32 subscriptionId = numberOfSubscriptions;
+        SubscriptionInfo storage sub = subscriptions[subscriptionId];
+        sub.subscriber = msg.sender;
+        paySubscriptionFor(subscriptionId);
+
+        subscribers[lookupKey(ritualId, msg.sender)] = subscriptionId;
+        numberOfSubscriptions += 1;
+
+        emit SubscriptionCreated(subscriptionId, msg.sender, ritualId);
+    }
+
+    /**
+     * @notice Pays for a subscription
+     * @param subscriptionId The ID of the subscription
+     */
+    function paySubscriptionFor(uint32 subscriptionId) public virtual {
+        uint256 amount = subscriptionFee();
+
+        SubscriptionInfo storage sub = subscriptions[subscriptionId];
+        sub.paidFor += amount;
+        sub.expiration += baseExpiration();
+
+        feeToken.safeTransferFrom(msg.sender, address(this), amount);
+
+        // TODO: We already emit SubscriptionCreated, do we need this?
+        emit SubscriptionPaid(subscriptionId, msg.sender, amount);
+    }
+
+    /**
+     * @notice Checks if a spender can spend from a subscription
+     * @param subscriptionId The ID of the subscription
+     * @param spender The address of the spender
+     * @return True if the spender can spend from the subscription, false otherwise
+     */
+    function canSpendFromSubscription(
+        // TODO: Currently unused, remove?
+        // solhint-disable-next-line no-unused-vars
+        uint32 subscriptionId,
+        address spender
+    ) public returns (bool) {
+        // By default, only coordinator can spend from subscription
+        return spender == address(coordinator);
+    }
+
+    /**
+     * @notice Spends from a subscription
+     * @param subscriptionId The ID of the subscription
+     * @param amount The amount to spend
+     */
+    function spendFromSubscription(uint32 subscriptionId, uint256 amount) external {
+        require(canSpendFromSubscription(subscriptionId, msg.sender), "Unauthorized spender");
+        feeToken.safeTransferFrom(address(this), msg.sender, amount);
+    }
+
+    /**
+     * @notice Withdraws the contract balance to the beneficiary
+     * @param amount The amount to withdraw
+     */
+    function withdrawToBeneficiary(uint256 amount) external {
+        require(msg.sender == beneficiary, "Only the beneficiary can withdraw");
+        uint256 contractBalance = feeToken.balanceOf(address(this));
+        require(contractBalance >= amount, "Insufficient contract balance");
```

Comment:
I think we can skip this check, because `feeToken` will do that anyway, unless here should be additional rules.
Other question here: if everything withdrawn to beneficiary (btw event would be great) then `cancelSubscription` would not work anymore, is it correct behavior?

Reply:
I think in that case we should only withdraw from expired subscriptions, to avoid breaking `cancelSubscription`.

Reply:
that makes sense, so I guess we want some check for that amount?